src/main/kotlin/com/pubnub/api/retry/RetryableBase.kt

```diff
@@ -0,0 +1,111 @@
+package com.pubnub.api.retry
+
+import com.pubnub.api.PubNubError
+import com.pubnub.api.PubNubException
+import retrofit2.Response
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+import javax.net.ssl.SSLHandshakeException
+import kotlin.math.pow
+
+internal abstract class RetryableBase<T>(
+    val retryPolicy: RequestRetryPolicy,
+    val endpointGroupName: RetryableEndpointGroup
+) {
+
+    companion object {
+        const val BOUND = 1000 // random delay should be between 0 and 1000 milliseconds
+        private const val TOO_MANY_REQUEST = 429
+        const val SERVICE_UNAVAILABLE = 503
+        private const val RETRY_AFTER_HEADER_NAME = "Retry-After"
+        const val MILLISECONDS = 1000
+        private val retryableStatusCodes = mapOf(
+            429 to "TOO_MANY_REQUESTS",
+            500 to "HTTP_INTERNAL_ERROR",
+            501 to "HTTP_NOT_IMPLEMENTED",
+            502 to "HTTP_BAD_GATEWAY",
+            503 to "HTTP_UNAVAILABLE",
+            504 to "HTTP_GATEWAY_TIMEOUT",
+            505 to "HTTP_VERSION",
+        )
+        internal val retryableExceptions = listOf(
+            UnknownHostException::class.java,
+            SocketTimeoutException::class.java,
+            ConnectException::class.java,
+            SSLHandshakeException::class.java,
+        )
+    }
+
+    private var exponentialMultiplier = 0.0
+
+    internal fun isRetryPolicySetForThisRestCall(): Boolean {
+        return when (retryPolicy) {
+            is RequestRetryPolicy.None -> false
+
+            is RequestRetryPolicy.Linear -> {
```

Comment:
I tried that but I got `Unresolved reference: excludedOperations`
in line: `val excludedOperations = retryPolicy.excludedOperations`

Both Linear and Exponential classes have the excludedOperations property, but the None object does not. Kotlin's smart casting cannot guarantee that excludedOperations exists on the retryPolicy object because it could potentially be RequestRetryPolicy.None.

Reply:
yeah, that's what I thought... let's just leave it, no point refactoring the class hierarchy just to do this.

Reply:
Yes, smart cast cannot access .excludedOperation. I have already tried this.